File: algorithm/DFS/DFS.java
import java.util.*;

public class dfs {
    static int[][] graph;
    static boolean[] visited;
    static int N;

    public static void dfs(int v) {
        visited[v] = true;
        System.out.print(v + " ");
        for (int i = 1; i <= N; i++) {
            if (graph[v][i] == 1 && !visited[i]) {
                dfs(i);
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        int M = sc.nextInt();
        int V = sc.nextInt();

        graph = new int[N + 1][N + 1];
        for (int i = 0; i < M; i++) {
            int a = sc.nextInt();
            int b = sc.nextInt();
            graph[a][b] = graph[b][a] = 1;
        }

        visited = new boolean[N + 1];
        dfs(V);
        sc.close();
    }
}

File: algorithm/DFS/DFS.py
# DFS (깊이 우선 탐색)
# 재귀를 이용해 깊은 노드부터 탐색

N, M, V = map(int, input().split())

graph = [[0]*(N+1) for _ in range(N+1)]
for _ in range(M):
    a, b = map(int, input().split())
    graph[a][b] = graph[b][a] = 1

visited = [0]*(N+1)

def dfs(V):
    visited[V] = 1
    print(V, end=' ')
    for i in range(1, N+1):
        if not visited[i] and graph[V][i] == 1:
            dfs(i)

dfs(V)
